File: server/app/api/plugins.py
"""插件与预设 API（DSH 组合理念：预设 = 提示词 + 工具白名单 + 技能白名单 + 模型覆盖）。"""

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session as DbSession

from app.db.database import get_db
from app.services import plugins as svc

router = APIRouter(prefix="/api/plugins", tags=["plugins"])


class ToolIn(BaseModel):
    name: str
    description: str = ""
    parameters: dict | None = None
    method: str = "GET"
    url: str
    headers: dict | None = None
    body_template: str = ""
    enabled: bool = True
    require_approval: bool = False


class ToolPatch(BaseModel):
    name: str | None = None
    description: str | None = None
    parameters: dict | None = None
    method: str | None = None
    url: str | None = None
    headers: dict | None = None
    body_template: str | None = None
    enabled: bool | None = None
    require_approval: bool | None = None


class PresetIn(BaseModel):
    name: str
    description: str = ""
    prompt_extra: str = ""
    tools: list[str] | None = None
    skills: list[str] | None = None
    model_override: str = ""
    enabled: bool = True


class PresetPatch(BaseModel):
    name: str | None = None
    description: str | None = None
    prompt_extra: str | None = None
    tools: list[str] | None = None
    skills: list[str] | None = None
    model_override: str | None = None
    enabled: bool | None = None


# ---------- 自定义工具（插件） ----------
@router.get("/tools")
def list_tools(db: DbSession = Depends(get_db)):
    return svc.list_custom_tools(db)


@router.post("/tools")
def create_tool(body: ToolIn, db: DbSession = Depends(get_db)):
    result = svc.create_custom_tool(db, **body.model_dump())
    if result.get("error"):
        raise HTTPException(400, result["error"])
    return result


@router.patch("/tools/{tool_id}")
def update_tool(tool_id: int, body: ToolPatch, db: DbSession = Depends(get_db)):
    result = svc.update_custom_tool(db, tool_id, **body.model_dump(exclude_none=True))
    if result.get("error"):
        raise HTTPException(404, result["error"])
    return result


@router.delete("/tools/{tool_id}")
def delete_tool(tool_id: int, db: DbSession = Depends(get_db)):
    result = svc.delete_custom_tool(db, tool_id)
    if result.get("error"):
        raise HTTPException(400, result["error"])
    return result


@router.post("/tools/{tool_id}/test")
def test_tool(tool_id: int, args: dict | None = None, db: DbSession = Depends(get_db)):
    from app.db.models import CustomTool

    row = db.get(CustomTool, tool_id)
    if row is None:
        raise HTTPException(404, f"工具不存在: {tool_id}")
    return svc.run_custom_tool(row, args or {})


# ---------- Agent 预设 ----------
@router.get("/presets")
def list_presets(db: DbSession = Depends(get_db)):
    return svc.list_presets(db)


@router.post("/presets")
def create_preset(body: PresetIn, db: DbSession = Depends(get_db)):
    result = svc.create_preset(db, **body.model_dump())
    if result.get("error"):
        raise HTTPException(400, result["error"])
    return result


@router.patch("/presets/{preset_id}")
def update_preset(preset_id: int, body: PresetPatch, db: DbSession = Depends(get_db)):
    result = svc.update_preset(db, preset_id, **body.model_dump(exclude_none=True))
    if result.get("error"):
        raise HTTPException(404, result["error"])
    return result


@router.delete("/presets/{preset_id}")
def delete_preset(preset_id: int, db: DbSession = Depends(get_db)):
    result = svc.delete_preset(db, preset_id)
    if result.get("error"):
        raise HTTPException(400, result["error"])
    return result


@router.post("/presets/seed")
def seed_presets(db: DbSession = Depends(get_db)):
    return {"created": svc.seed_builtin_presets(db)}

File: backend/src/api/plugins.rs
//! 插件（手搓 HTTP 工具）与 Agent 预设 API：`/api/plugins/*`。

use axum::extract::{Path, State};
use axum::Json;
use serde::Deserialize;
use serde_json::{json, Value};

use crate::error::{AppError, AppResult};
use crate::services::presets;
use crate::state::AppState;

// ---------- 插件（自定义工具） ----------

/// GET /api/plugins/tools
pub async fn list_tools(State(st): State<AppState>) -> AppResult<Json<Value>> {
    Ok(Json(presets::list_custom_tools(&st.db)?))
}

/// POST /api/plugins/tools
pub async fn create_tool(
    State(st): State<AppState>,
    Json(body): Json<Value>,
) -> AppResult<Json<Value>> {
    let result = presets::create_custom_tool(&st.db, &body)?;
    if let Some(err) = result.get("error").and_then(|v| v.as_str()) {
        return Err(AppError::bad_request(err));
    }
    Ok(Json(result))
}

/// PATCH /api/plugins/tools/{id}
pub async fn update_tool(
    State(st): State<AppState>,
    Path(id): Path<i64>,
    Json(body): Json<Value>,
) -> AppResult<Json<Value>> {
    let result = presets::update_custom_tool(&st.db, id, &body)?;
    if let Some(err) = result.get("error").and_then(|v| v.as_str()) {
        return Err(AppError::not_found(err));
    }
    Ok(Json(result))
}

/// DELETE /api/plugins/tools/{id}
pub async fn delete_tool(State(st): State<AppState>, Path(id): Path<i64>) -> AppResult<Json<Value>> {
    let result = presets::delete_custom_tool(&st.db, id)?;
    if let Some(err) = result.get("error").and_then(|v| v.as_str()) {
        return Err(AppError::bad_request(err));
    }
    Ok(Json(result))
}

/// POST /api/plugins/tools/{id}/test —— 用给的参数实调一次（设置面板"测试"按钮）
pub async fn test_tool(
    State(st): State<AppState>,
    Path(id): Path<i64>,
    body: Option<Json<Value>>,
) -> AppResult<Json<Value>> {
    let row = presets::custom_tool_by_id(&st.db, id)?
        .ok_or_else(|| AppError::not_found(format!("工具不存在: {id}")))?;
    let args = body.map(|Json(v)| v).unwrap_or_else(|| json!({}));
    Ok(Json(presets::run_custom_tool(&row, &args).await))
}

// ---------- Agent 预设 ----------

/// GET /api/plugins/presets
pub async fn list_presets(State(st): State<AppState>) -> AppResult<Json<Value>> {
    Ok(Json(presets::list_presets(&st.db)?))
}

/// POST /api/plugins/presets
pub async fn create_preset(
    State(st): State<AppState>,
    Json(body): Json<Value>,
) -> AppResult<Json<Value>> {
    let result = presets::create_preset(&st.db, &body)?;
    if let Some(err) = result.get("error").and_then(|v| v.as_str()) {
        return Err(AppError::bad_request(err));
    }
    Ok(Json(result))
}

/// PATCH /api/plugins/presets/{id}
pub async fn update_preset(
    State(st): State<AppState>,
    Path(id): Path<i64>,
    Json(body): Json<Value>,
) -> AppResult<Json<Value>> {
    let result = presets::update_preset(&st.db, id, &body)?;
    if let Some(err) = result.get("error").and_then(|v| v.as_str()) {
        return Err(AppError::not_found(err));
    }
    Ok(Json(result))
}

/// DELETE /api/plugins/presets/{id}
pub async fn delete_preset(
    State(st): State<AppState>,
    Path(id): Path<i64>,
) -> AppResult<Json<Value>> {
    let result = presets::delete_preset(&st.db, id)?;
    if let Some(err) = result.get("error").and_then(|v| v.as_str()) {
        return Err(AppError::bad_request(err));
    }
    Ok(Json(result))
}

/// POST /api/plugins/presets/seed —— 补齐内置预设（幂等）
pub async fn seed_presets(State(st): State<AppState>) -> AppResult<Json<Value>> {
    let created = presets::seed_builtin_presets(&st.db)?;
    Ok(Json(json!({ "created": created })))
}

/// GET /api/plugins/presets/{id}/export
pub async fn export_preset(
    State(st): State<AppState>,
    Path(id): Path<i64>,
) -> AppResult<Json<Value>> {
    let result = presets::export_preset(&st.db, id)?;
    if let Some(err) = result.get("error").and_then(|v| v.as_str()) {
        return Err(AppError::not_found(err));
    }
    Ok(Json(result))
}

/// GET /api/plugins/export —— 导出全部预设 + 技能 + 插件
pub async fn export_all(State(st): State<AppState>) -> AppResult<Json<Value>> {
    Ok(Json(presets::export_all(&st.db)?))
}

#[derive(Deserialize)]
pub struct ImportReq {
    pub data: Value,
    #[serde(default = "default_strategy")]
    pub strategy: String,
}

fn default_strategy() -> String {
    "rename".into()
}

/// POST /api/plugins/import
pub async fn import_bundle(
    State(st): State<AppState>,
    Json(req): Json<ImportReq>,
) -> AppResult<Json<Value>> {
    let result = presets::import_bundle(&st.db, &req.data, &req.strategy)?;
    if let Some(err) = result.get("error").and_then(|v| v.as_str()) {
        return Err(AppError::bad_request(err));
    }
    Ok(Json(result))
}
